src/components/Game/Game.jsx

```diff
@@ -0,0 +1,162 @@
+import {History} from "../History/History";
+import '../../App.css'
+import {GameMenu} from "../GameMenu/GameMenu";
+import {GameInfo} from "../GameInfo/GameInfo";
+import {GAME_SYMBOLS} from "../../scripts/libraries";
+import {bestBotMove} from "../../scripts/bot-move";
+
+
+class Game extends React.Component {
+    constructor(props) {
+        super(props);
+        this.state = this.initStartState();
+        this.jumpToHandleClick = this.jumpTo.bind(this);
+        this.boardHandleClick = this.handleClick.bind(this);
+        this.startNewGameHandler = this.startNewGame.bind(this);
+        this.enableBotModeHandler = this.enableBotMode.bind(this);
+        this.disableBotModeHandler = this.disableBotMode.bind(this)
+        this.resetScoreHandler = this.resetScore.bind(this);
+    }
+
+    componentDidMount() {
+        const data = getLocalStorageData(LocalStorageKeys.game);
+        if (data) {
+            this.setState(data);
+        }
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        if (!this.state.isGameOver) {
+            const lastSquares = this.state.history[this.state.history.length - 1].squares;
+            const winner = calculateWinner(lastSquares);
+            if (winner) {
+                this.setState({
+                    isGameOver: true,
+                    score: {
+                        winsX: winner === GAME_SYMBOLS.X ? this.state.score.winsX + 1 : this.state.score.winsX,
+                        winsO: winner === GAME_SYMBOLS.O ? this.state.score.winsO + 1 : this.state.score.winsO,
+                    }
+                })
+            } else if (lastSquares.indexOf(null) < 0) {
+                this.setState({
+                    isGameOver: true,
+                })
+            }
+        }
+        if (prevState !== this.state) {
+            updateLocalStorageData(LocalStorageKeys.game, this.state);
+        }
+    }
+
+    initStartState() {
+        return {
+            history: [{
+                squares: Array(9).fill(null),
+            }],
+            xIsNext: true,
+            stepNumber: 0,
+            botMode: false,
+            isGameOver: false,
+            isMoveBlocked: false,
+            score: {
+                winsX: 0,
+                winsO: 0,
+            }
+        }
+    }
+
+    fillSquare(i) {
+        const history = this.state.history.slice(0, this.state.stepNumber + 1);
+        const currentSquares = history[history.length - 1].squares;
+        if (!(calculateWinner(currentSquares) || currentSquares[i])) {
+            const squares = [...currentSquares];
+            squares[i] = this.state.xIsNext ? GAME_SYMBOLS.X : GAME_SYMBOLS.O;
+            this.setState(() => {
+                return {
+                    history: history.concat([{squares: squares}]),
+                    xIsNext: !this.state.xIsNext,
+                    stepNumber: history.length,
+                }
```

Comment:
you can do all you changes right inside setState callback function, you also have to spread(merge) previous state with new state `this.setState((prevState) => ({ ...prevState, history: [...prevState.history, squares ], xIsNext: !prevState.next, ... }))` and `it's very important!!` you shouldn't use `this` inside `setState`, because you have access to `previous state`!

Reply:
```suggestion
           this.setState((prevState) => {
                return {
                    ...prevState,
                    history: history.concat([{squares: squares}]),
                    xIsNext: !prevState.xIsNext,
                    stepNumber: history.length,
                }
            });
```

Reply:
```suggestion
fillSquare(i) {
        const history = getHistory(this.state);
        const currentSquares = history[history.length - 1].squares;
        if (!(calculateWinner(currentSquares) || currentSquares[i])) {
            this.setState((prevState) => {
                const prevHistory = getHistory(prevState);
                const squares = [...prevHistory[prevHistory.length - 1].squares];
                squares[i] = this.state.xIsNext ? GAME_SYMBOLS.X : GAME_SYMBOLS.O;
                return {
                    ...prevState,
                    history: prevHistory.concat([{squares}]),
                    xIsNext: !prevState.xIsNext,
                    stepNumber: prevHistory.length,
                }
            });
        }
    }
```
Fixed